File: src/plugins/persistence/fake/repositories.py
"""Fake in-memory repositories.

Each repository keeps its records in a mutable in-memory store. One shared store
object can be handed to several repositories so a ``FakeUnitOfWork`` transaction
sees consistent data across conversations / envelopes / sessions / assets.

Entity ``id`` values are stable UUIDs generated at creation time (via
``new_id``), so repositories simply store the entities as given.
"""

from __future__ import annotations

import asyncio
from dataclasses import replace

from src.core.entities.agent_run import AgentRun
from src.core.entities.outbox_event import OutboxEvent
from src.core.entities.conversation import Conversation
from src.core.entities.message import Message
from src.core.entities.session import Session
from src.core.entities.session_envelope import SessionEnvelope
from src.core.entities.transport.address import ConversationAddress
from src.core.entities.transport.asset import Asset
from src.core.entities.transport.envelope import MessageEnvelope
from src.core.values import SessionEnvelopeRole, SessionStatus, now_ms

__all__ = [
    "FakeOutboxRepository",
    "MemoryStore",
    "FakeConversationRepository",
    "FakeEnvelopeRepository",
    "FakeMessageRepository",
    "FakeSessionRepository",
    "FakeSessionEnvelopeRepository",
    "FakeAssetRepository",
    "FakeAgentRunRepository",
]


class MemoryStore:
    """Shared in-memory backing store used by all fake repositories."""

    def __init__(self) -> None:
        self.transaction_lock = asyncio.Lock()
        self.conversations: dict[str, Conversation] = {}
        self.envelopes: dict[str, MessageEnvelope] = {}
        self.messages: dict[str, Message] = {}
        self.sessions: dict[str, Session] = {}
        self.session_envelopes: dict[tuple[str, str], SessionEnvelope] = {}
        self.assets: list[Asset] = []
        self.agent_runs: dict[str, AgentRun] = {}
        self.outbox: dict[str, OutboxEvent] = {}


class FakeConversationRepository:
    """``ConversationRepository`` stored in memory, keyed by address."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    @staticmethod
    def _key(address: ConversationAddress) -> str:
        return address.identity_key()

    async def add(self, conversation: Conversation) -> None:
        if conversation.address is not None:
            self._store.conversations[self._key(conversation.address)] = conversation

    async def get(self, conversation_id: str) -> Conversation | None:
        for c in self._store.conversations.values():
            if str(c.id) == conversation_id:
                return c
        return None

    async def get_or_create_by_address(
        self, address: ConversationAddress
    ) -> Conversation:
        key = self._key(address)
        existing = self._store.conversations.get(key)
        if existing is not None:
            return existing
        conversation = Conversation(address=address)
        self._store.conversations[key] = conversation
        return conversation

    async def save(self, conversation: Conversation) -> None:
        if conversation.address is not None:
            self._store.conversations[self._key(conversation.address)] = conversation


class FakeEnvelopeRepository:
    """``EnvelopeRepository`` stored in memory, keyed by envelope id."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, envelope: MessageEnvelope) -> None:
        if envelope.conversation_id is None:
            raise ValueError("envelope conversation_id is required")
        if envelope.message is not None:
            self._store.messages[envelope.message.id] = envelope.message
        self._store.envelopes[envelope.id] = envelope

    async def get(self, envelope_id: str) -> MessageEnvelope | None:
        return self._store.envelopes.get(envelope_id)

    async def get_by_idempotency_key(
        self, conversation_id: str, idempotency_key: str
    ) -> MessageEnvelope | None:
        return next((e for e in self._store.envelopes.values()
                     if e.conversation_id == conversation_id
                     and e.idempotency_key == idempotency_key), None)

    async def get_by_external_message_id(
        self,
        *,
        conversation_id: str,
        external_message_id: str,
    ) -> MessageEnvelope | None:
        for envelope in self._store.envelopes.values():
            if envelope.conversation_id != conversation_id:
                continue
            if (
                envelope.transport is not None
                and envelope.transport.external_message_id == external_message_id
            ):
                return envelope
        return None

    async def list_by_conversation(
        self,
        conversation_id: str,
        *,
        after_envelope_id: str | None = None,
        limit: int | None = None,
    ) -> list[MessageEnvelope]:
        items = [
            e for e in self._store.envelopes.values()
            if e.conversation_id == conversation_id
        ]
        items.sort(key=lambda e: str(e.id or ""))
        if after_envelope_id is not None:
            seen = False
            filtered: list[MessageEnvelope] = []
            for e in items:
                if seen:
                    filtered.append(e)
                elif str(e.id) == after_envelope_id:
                    seen = True
            items = filtered
        if limit is not None:
            items = items[-limit:]
        return items


class FakeOutboxRepository:
    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, event: OutboxEvent) -> None:
        if event.run_id not in self._store.agent_runs:
            raise ValueError("outbox run does not exist")
        if event.envelope_id not in self._store.envelopes:
            raise ValueError("outbox envelope does not exist")
        if any(e.run_id == event.run_id or e.envelope_id == event.envelope_id
               for e in self._store.outbox.values()):
            raise ValueError("duplicate execution event")
        self._store.outbox[event.id] = event

    async def get_by_envelope(self, envelope_id: str) -> OutboxEvent | None:
        return next((e for e in self._store.outbox.values()
                     if e.envelope_id == envelope_id), None)

    async def list_pending(self, limit: int = 100) -> list[OutboxEvent]:
        if limit < 1:
            raise ValueError("limit must be positive")
        return sorted((e for e in self._store.outbox.values()
                       if e.published_at is None),
                      key=lambda e: (e.created_at, e.id))[:limit]

    async def mark_published(self, event_id: str, published_at: int) -> None:
        event = self._store.outbox[event_id]
        if event.published_at is None:
            self._store.outbox[event_id] = replace(event, published_at=published_at)


class FakeMessageRepository:
    """``MessageRepository`` stored in memory, keyed by message id."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, message: Message) -> None:
        self._store.messages[message.id] = message

    async def get(self, message_id: str) -> Message | None:
        return self._store.messages.get(message_id)


class FakeSessionRepository:
    """``SessionRepository`` stored in memory, keyed by session id."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, session: Session) -> None:
        self._store.sessions[session.id] = session

    async def get(self, session_id: str) -> Session | None:
        return self._store.sessions.get(session_id)

    async def save(self, session: Session) -> None:
        if session.id is not None:
            self._store.sessions[session.id] = session

    async def get_active(
        self,
        *,
        conversation_id: str,
        owner_user_id: str | None = None,
    ) -> Session | None:
        for s in self._store.sessions.values():
            if s.conversation_id != conversation_id:
                continue
            if owner_user_id is not None and s.owner_user_id != owner_user_id:
                continue
            if s.status in (
                SessionStatus.QUEUED,
                SessionStatus.RUNNING,
                SessionStatus.WAITING_USER,
            ):
                return s
        return None

    async def list_by_conversation(self, conversation_id: str) -> list[Session]:
        return [s for s in self._store.sessions.values() if s.conversation_id == conversation_id]


class FakeAssetRepository:
    """``AssetRepository`` stored in memory."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, asset: Asset) -> None:
        self._store.assets.append(asset)

    async def get(self, asset_id: str) -> Asset | None:
        for a in self._store.assets:
            if str(a.id) == asset_id:
                return a
        return None

    async def list_by_sha256(self, sha256: str) -> list[Asset]:
        return [a for a in self._store.assets if a.sha256 == sha256]


class FakeSessionEnvelopeRepository:
    """Ordered, deduplicated Session-to-Envelope relations in memory."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def attach(
        self,
        *,
        session_id: str,
        envelope_id: str,
        relation_role: SessionEnvelopeRole = SessionEnvelopeRole.INPUT,
    ) -> SessionEnvelope:
        key = (session_id, envelope_id)
        existing = self._store.session_envelopes.get(key)
        if existing is not None:
            return existing

        session = self._store.sessions.get(session_id)
        envelope = self._store.envelopes.get(envelope_id)
        if session is None:
            raise ValueError(f"unknown session: {session_id}")
        if envelope is None:
            raise ValueError(f"unknown envelope: {envelope_id}")
        if session.conversation_id != envelope.conversation_id:
            raise ValueError("session and envelope belong to different conversations")

        sequence_no = 1 + max(
            (
                link.sequence_no
                for link in self._store.session_envelopes.values()
                if link.session_id == session_id
            ),
            default=0,
        )
        link = SessionEnvelope(
            session_id=session_id,
            envelope_id=envelope_id,
            sequence_no=sequence_no,
            relation_role=relation_role,
            created_at=now_ms(),
        )
        self._store.session_envelopes[key] = link
        return link

    async def list_by_session(self, session_id: str) -> list[SessionEnvelope]:
        links = [
            link
            for link in self._store.session_envelopes.values()
            if link.session_id == session_id
        ]
        return sorted(links, key=lambda link: link.sequence_no)

    async def list_session_ids(self, envelope_id: str) -> list[str]:
        links = [
            link
            for link in self._store.session_envelopes.values()
            if link.envelope_id == envelope_id
        ]
        links.sort(key=lambda link: (link.created_at or 0, link.session_id))
        return [link.session_id for link in links]


class FakeAgentRunRepository:
    """``AgentRunRepository`` stored in memory."""

    def __init__(self, store: MemoryStore) -> None:
        self._store = store

    async def add(self, run: AgentRun) -> None:
        if run.session_id is None or run.session_id not in self._store.sessions:
            raise ValueError(f"unknown session: {run.session_id}")
        self._store.agent_runs[run.id] = run

    async def get(self, run_id: str) -> AgentRun | None:
        return self._store.agent_runs.get(run_id)

    async def save(self, run: AgentRun) -> None:
        self._store.agent_runs[run.id] = run

    async def list_by_session(self, session_id: str) -> list[AgentRun]:
        runs = [
            run for run in self._store.agent_runs.values()
            if run.session_id == session_id
        ]
        return sorted(runs, key=lambda run: (run.attempt, run.id))

File: packages/persistence/src/memory/repositories.ts
import {
  Conversation,
  SessionEnvelope,
  SessionEnvelopeRole,
  type AgentRun,
  type Asset,
  type ConversationAddress,
  type EntityId,
  type Message,
  type MessageEnvelope,
  type OutboxEvent,
  type Session,
  type UnixMillis,
  nowMs,
} from "@kookbot/domain";
import type {
  AgentRunRepository,
  AssetRepository,
  ConversationRepository,
  EnvelopeRepository,
  MessageRepository,
  OutboxRepository,
  SessionEnvelopeRepository,
  SessionRepository,
} from "@kookbot/application";

import type { MemoryState } from "./store.js";

function required<T>(value: T | undefined, message: string): T {
  if (value === undefined) throw new TypeError(message);
  return value;
}

export class MemoryConversationRepository implements ConversationRepository {
  constructor(private readonly state: MemoryState) {}

  async add(conversation: Conversation): Promise<void> {
    if (this.state.conversations.has(conversation.id))
      throw new Error(`conversation ${conversation.id} already exists`);
    const address = required(conversation.address, "conversation address is required");
    const key = address.identityKey();
    if (this.state.conversationIdsByAddress.has(key)) throw new Error("conversation address already exists");
    this.state.conversations.set(conversation.id, conversation.clone());
    this.state.conversationIdsByAddress.set(key, conversation.id);
  }

  async get(conversationId: EntityId): Promise<Conversation | undefined> {
    return this.state.conversations.get(conversationId)?.clone();
  }

  async getOrCreateByAddress(address: ConversationAddress): Promise<Conversation> {
    const key = address.identityKey();
    const existingId = this.state.conversationIdsByAddress.get(key);
    if (existingId)
      return required(this.state.conversations.get(existingId), "conversation address index is corrupt").clone();
    const conversation = new Conversation({ address });
    await this.add(conversation);
    return conversation.clone();
  }

  async save(conversation: Conversation): Promise<void> {
    if (!this.state.conversations.has(conversation.id))
      throw new Error(`conversation ${conversation.id} does not exist`);
    const previous = required(this.state.conversations.get(conversation.id), "conversation is missing");
    const previousKey = previous.address?.identityKey();
    const nextAddress = required(conversation.address, "conversation address is required");
    const nextKey = nextAddress.identityKey();
    const owner = this.state.conversationIdsByAddress.get(nextKey);
    if (owner !== undefined && owner !== conversation.id) throw new Error("conversation address already exists");
    if (previousKey !== undefined && previousKey !== nextKey) this.state.conversationIdsByAddress.delete(previousKey);
    this.state.conversationIdsByAddress.set(nextKey, conversation.id);
    this.state.conversations.set(conversation.id, conversation.clone());
  }
}

export class MemoryMessageRepository implements MessageRepository {
  constructor(private readonly state: MemoryState) {}
  async add(message: Message): Promise<void> {
    if (this.state.messages.has(message.id)) throw new Error(`message ${message.id} already exists`);
    this.state.messages.set(message.id, message.clone());
  }
  async get(messageId: EntityId): Promise<Message | undefined> {
    return this.state.messages.get(messageId)?.clone();
  }
}

export class MemoryEnvelopeRepository implements EnvelopeRepository {
  constructor(private readonly state: MemoryState) {}

  async add(envelope: MessageEnvelope): Promise<void> {
    if (this.state.envelopes.has(envelope.id)) throw new Error(`envelope ${envelope.id} already exists`);
    const conversationId = required(envelope.conversationId, "envelope conversationId is required");
    if (envelope.idempotencyKey && (await this.getByIdempotencyKey(conversationId, envelope.idempotencyKey))) {
      throw new Error("envelope idempotency key already exists");
    }
    const externalMessageId = envelope.transport?.externalMessageId;
    if (externalMessageId && (await this.getByExternalMessageId({ conversationId, externalMessageId }))) {
      throw new Error("external message id already exists");
    }
    this.state.envelopes.set(envelope.id, envelope.clone());
  }

  async get(envelopeId: EntityId): Promise<MessageEnvelope | undefined> {
    return this.state.envelopes.get(envelopeId)?.clone();
  }

  async getByIdempotencyKey(conversationId: EntityId, idempotencyKey: string): Promise<MessageEnvelope | undefined> {
    for (const envelope of this.state.envelopes.values()) {
      if (envelope.conversationId === conversationId && envelope.idempotencyKey === idempotencyKey)
        return envelope.clone();
    }
    return undefined;
  }

  async getByExternalMessageId(input: {
    readonly conversationId: EntityId;
    readonly externalMessageId: string;
  }): Promise<MessageEnvelope | undefined> {
    for (const envelope of this.state.envelopes.values()) {
      if (
        envelope.conversationId === input.conversationId &&
        envelope.transport?.externalMessageId === input.externalMessageId
      ) {
        return envelope.clone();
      }
    }
    return undefined;
  }

  async listByConversation(
    conversationId: EntityId,
    options: { readonly afterEnvelopeId?: EntityId; readonly limit?: number } = {},
  ): Promise<MessageEnvelope[]> {
    let ordered = [...this.state.envelopes.values()]
      .filter((item) => item.conversationId === conversationId)
      .sort((left, right) => left.id.localeCompare(right.id));
    if (options.afterEnvelopeId !== undefined) {
      const afterIndex = ordered.findIndex((item) => item.id === options.afterEnvelopeId);
      ordered = afterIndex < 0 ? [] : ordered.slice(afterIndex + 1);
    }
    if (options.limit !== undefined) ordered = ordered.slice(-options.limit);
    return ordered.map((item) => item.clone());
  }
}

export class MemorySessionRepository implements SessionRepository {
  constructor(private readonly state: MemoryState) {}
  async add(session: Session): Promise<void> {
    if (this.state.sessions.has(session.id)) throw new Error(`session ${session.id} already exists`);
    this.state.sessions.set(session.id, session.clone());
  }
  async get(sessionId: EntityId): Promise<Session | undefined> {
    return this.state.sessions.get(sessionId)?.clone();
  }
  async save(session: Session): Promise<void> {
    if (!this.state.sessions.has(session.id)) throw new Error(`session ${session.id} does not exist`);
    this.state.sessions.set(session.id, session.clone());
  }
  async getActive(input: {
    readonly conversationId: EntityId;
    readonly ownerUserId?: string;
  }): Promise<Session | undefined> {
    return [...this.state.sessions.values()]
      .filter(
        (session) =>
          session.conversationId === input.conversationId &&
          (input.ownerUserId === undefined || session.ownerUserId === input.ownerUserId) &&
          session.isActive(),
      )
      .sort((left, right) => (right.updatedAt ?? 0) - (left.updatedAt ?? 0))[0]
      ?.clone();
  }
  async listByConversation(conversationId: EntityId): Promise<Session[]> {
    return [...this.state.sessions.values()]
      .filter((session) => session.conversationId === conversationId)
      .map((session) => session.clone());
  }
}

export class MemorySessionEnvelopeRepository implements SessionEnvelopeRepository {
  constructor(private readonly state: MemoryState) {}
  async attach(input: {
    readonly sessionId: EntityId;
    readonly envelopeId: EntityId;
    readonly relationRole?: (typeof SessionEnvelopeRole)[keyof typeof SessionEnvelopeRole];
  }): Promise<SessionEnvelope> {
    const session = required(this.state.sessions.get(input.sessionId), `session ${input.sessionId} does not exist`);
    const envelope = required(
      this.state.envelopes.get(input.envelopeId),
      `envelope ${input.envelopeId} does not exist`,
    );
    if (session.conversationId !== envelope.conversationId) {
      throw new TypeError("session and envelope belong to different conversations");
    }
    const key = `${input.sessionId}\0${input.envelopeId}`;
    const existing = this.state.sessionEnvelopes.get(key);
    if (existing) return existing.clone();
    const sequenceNo =
      Math.max(
        0,
        ...[...this.state.sessionEnvelopes.values()]
          .filter((relation) => relation.sessionId === input.sessionId)
          .map((relation) => relation.sequenceNo),
      ) + 1;
    const relation = new SessionEnvelope({
      sessionId: input.sessionId,
      envelopeId: input.envelopeId,
      sequenceNo,
      relationRole: input.relationRole ?? SessionEnvelopeRole.INPUT,
      createdAt: nowMs(),
    });
    this.state.sessionEnvelopes.set(key, relation.clone());
    return relation.clone();
  }
  async listBySession(sessionId: EntityId): Promise<SessionEnvelope[]> {
    return [...this.state.sessionEnvelopes.values()]
      .filter((relation) => relation.sessionId === sessionId)
      .sort((left, right) => left.sequenceNo - right.sequenceNo)
      .map((relation) => relation.clone());
  }
  async listSessionIds(envelopeId: EntityId): Promise<EntityId[]> {
    return [...this.state.sessionEnvelopes.values()]
      .filter((relation) => relation.envelopeId === envelopeId)
      .map((relation) => relation.sessionId);
  }
}

export class MemoryAgentRunRepository implements AgentRunRepository {
  constructor(private readonly state: MemoryState) {}
  async add(run: AgentRun): Promise<void> {
    if (this.state.agentRuns.has(run.id)) throw new Error(`agent run ${run.id} already exists`);
    if (!run.sessionId || !this.state.sessions.has(run.sessionId)) {
      throw new TypeError(`unknown session: ${run.sessionId ?? "undefined"}`);
    }
    this.state.agentRuns.set(run.id, run.clone());
  }
  async get(runId: EntityId): Promise<AgentRun | undefined> {
    return this.state.agentRuns.get(runId)?.clone();
  }
  async save(run: AgentRun): Promise<void> {
    if (!this.state.agentRuns.has(run.id)) throw new Error(`agent run ${run.id} does not exist`);
    this.state.agentRuns.set(run.id, run.clone());
  }
  async listBySession(sessionId: EntityId): Promise<AgentRun[]> {
    return [...this.state.agentRuns.values()]
      .filter((run) => run.sessionId === sessionId)
      .sort((left, right) => left.attempt - right.attempt)
      .map((run) => run.clone());
  }
}

export class MemoryAssetRepository implements AssetRepository {
  constructor(private readonly state: MemoryState) {}
  async add(asset: Asset): Promise<void> {
    if (this.state.assets.has(asset.id)) throw new Error(`asset ${asset.id} already exists`);
    this.state.assets.set(asset.id, asset.clone());
  }
  async get(assetId: EntityId): Promise<Asset | undefined> {
    return this.state.assets.get(assetId)?.clone();
  }
  async listBySha256(sha256: string): Promise<Asset[]> {
    return [...this.state.assets.values()].filter((asset) => asset.sha256 === sha256).map((asset) => asset.clone());
  }
}

export class MemoryOutboxRepository implements OutboxRepository {
  constructor(private readonly state: MemoryState) {}
  async add(event: OutboxEvent): Promise<void> {
    if (this.state.outboxEvents.has(event.id)) throw new Error(`outbox event ${event.id} already exists`);
    if (!this.state.agentRuns.has(event.runId)) throw new TypeError(`unknown agent run: ${event.runId}`);
    if (!this.state.envelopes.has(event.envelopeId)) throw new TypeError(`unknown envelope: ${event.envelopeId}`);
    if (
      [...this.state.outboxEvents.values()].some(
        (item) => item.runId === event.runId || item.envelopeId === event.envelopeId,
      )
    ) {
      throw new Error("duplicate execution event");
    }
    this.state.outboxEvents.set(event.id, event.clone());
  }
  async getByEnvelope(envelopeId: EntityId): Promise<OutboxEvent | undefined> {
    return [...this.state.outboxEvents.values()].find((event) => event.envelopeId === envelopeId)?.clone();
  }
  async listPending(limit = 100): Promise<OutboxEvent[]> {
    if (!Number.isSafeInteger(limit) || limit < 1) throw new TypeError("limit must be positive");
    return [...this.state.outboxEvents.values()]
      .filter((event) => event.publishedAt === undefined)
      .sort((left, right) => left.createdAt - right.createdAt || left.id.localeCompare(right.id))
      .slice(0, limit)
      .map((event) => event.clone());
  }
  async markPublished(eventId: EntityId, publishedAt: UnixMillis): Promise<void> {
    const event = required(this.state.outboxEvents.get(eventId), `outbox event ${eventId} does not exist`);
    if (event.publishedAt === undefined) this.state.outboxEvents.set(eventId, event.withPublishedAt(publishedAt));
  }
}
